lib/aoch.ex

```diff
@@ -0,0 +1,229 @@
+defmodule AoCH do
+  @moduledoc """
+  AoCH keeps the contexts that define your domain
+  and business logic.
+
+  Contexts are also responsible for managing your data, regardless
+  if it comes from the database, an external API or others.
+  """
+
+  def get_leaderboard_day(day, year) do
+    {leaderboard_day, _, _} = get_data(year, day)
+    leaderboard_day
+  end
+
+  def get_leaderboard_year(day, year) do
+    {_, leaderboard, _} = get_data(year, day)
+    leaderboard
+  end
+
+  def get_api_data(day, year) do
+    {_, _, data} = get_data(year, day)
+    data
+  end
+
+  def get_challenge(day, year) do
+    ConCache.get_or_store(:cache, "challenge_#{year}_#{day}", fn ->
+      %ConCache.Item{value: request_day_challenge(day, year), ttl: :infinity}
+    end)
+  end
+
+  def now() do
+    DateTime.utc_now() |> DateTime.shift_zone!("America/New_York")
+  end
+
+  def start_of_day(year, day) do
+    now = now()
+
+    %DateTime{
+      now
+      | year: year,
+        month: 12,
+        day: day,
+        hour: 0,
+        minute: 0,
+        second: 0,
+        microsecond: {0, 0}
+    }
+  end
+
+  defp get_data(year, day) do
+    ConCache.get_or_store(:cache, "data_#{year}", fn ->
+      raw_data = request_raw_data(year)
+      leaderboard_day = parse_leaderboard_day(raw_data, day, year)
+      leaderboard = parse_leaderboard_year(raw_data)
+      day_challenge = get_challenge(day, year)
+
+      data =
+        %{
+          assignment: day_challenge,
+          today: leaderboard_day |> Enum.map(&assign_time_strings/1),
+          total: leaderboard
+        }
+
+      {leaderboard_day, leaderboard, data}
+    end)
+  end
+
+  defp assign_time_strings(map) do
+    map =
+      if map[:first_star] do
+        Map.put(map, :star1, AoCHWeb.format_time(map[:first_star]))
+      else
+        map
+      end
+
+    if map[:second_star] do
+      Map.put(map, :star2, "+ " <> AoCHWeb.format_time(map[:second_star]))
+    else
+      map
+    end
```

Comment:
Very much nagging, however would it be possible for this section to `null` the `star2` field if the first is submitted but the second is not yet (like with the `second_star` field)?

example:
![image](https://github.com/user-attachments/assets/db5e3792-de7b-4b7e-a4d6-16da2eee2816)

*possible suggestion, however never used elixir/phoenix so please do not blindly accept*
```suggestion
    map =
      if map[:first_star] do
        Map.put(map, :star1, AoCHWeb.format_time(map[:first_star]))
      else
        map
      end

    if map[:second_star] do
      Map.put(map, :star2, "+ " <> AoCHWeb.format_time(map[:second_star]))
    else
      Map.put(map, :star2, nil)
    end
```

Reply:
Yes that ofcourse can be done. However, I did it like this on purpose to stay consistent with the old api.

Reply:
That is true, i'll leave it as is for now and if someone really wants/would like that to change they can create a PR/issue for it.